File: dallosh/clients/dallosh_web/src/lib/sodular/api/base-client.ts
import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
import { io, Socket } from 'socket.io-client';
import { ApiResponse, AuthTokens } from '../types/schema';
import { buildQueryParams, buildApiUrl, storage, TOKEN_KEYS } from '../utils';


export interface SodularClientConfig {
  baseUrl: string;
  timeout?: number;
}

export class BaseClient {
  public axiosInstance: AxiosInstance;
  private baseUrl: string;
  private currentDatabaseId?: string;
  private accessToken?: string;
  private refreshToken?: string;
  private isRefreshing = false;
  private refreshPromise: Promise<boolean> | null = null;
  private socket: Socket | null = null;

  constructor(config: SodularClientConfig) {
    this.baseUrl = config.baseUrl;
    
    this.axiosInstance = axios.create({
      baseURL: config.baseUrl,
      timeout: config.timeout || 30000,
      headers: {
        'Content-Type': 'application/json',
      },
    });

    this.loadTokensFromStorage();
    
    this.axiosInstance.interceptors.request.use((config) => {
      if (this.accessToken) {
        config.headers.Authorization = `Bearer ${this.accessToken}`;
      }
      return config;
    });

    this.axiosInstance.interceptors.response.use(
      (response) => response,
      async (error) => {
        const originalRequest = error.config;
        if (error.response?.status === 401 && !originalRequest._retry) {
          originalRequest._retry = true;
          
          if (!this.isRefreshing) {
            this.refreshPromise = this.performTokenRefresh();
          }

          const refreshed = await this.refreshPromise;
          
          if (refreshed) {
            originalRequest.headers.Authorization = `Bearer ${this.accessToken}`;
            return this.axiosInstance(originalRequest);
          }
        }
        return Promise.reject(error);
      }
    );
  }

  async connect(): Promise<{ isReady: boolean; error?: string }> {
    try {
      await this.axiosInstance.get('/health');
      
      // Connect to socket server
      this.connectSocket();
      
      return { isReady: true };
    } catch (error: any) {
      return { 
        isReady: false, 
        error: error.message || 'Failed to connect to Sodular backend',
      };
    }
  }

  /**
   * Extract host URL from API base URL
   * e.g., 'http://sodular_backend:5001/api/v1' -> 'http://sodular_backend:5001'
   */
  private getSocketUrl(): string {
    try {
      const url = new URL(this.baseUrl);
      return `${url.protocol}//${url.host}`;
    } catch (error) {
      console.error('Failed to parse base URL for socket connection:', error);
      // Fallback: try to remove common API paths
      return this.baseUrl
        .replace(/\/api\/v\d+\/?$/, '')
        .replace(/\/api\/?$/, '')
        .replace(/\/$/, '');
    }
  }

  /**
   * Connect to socket server
   */
  private connectSocket(): void {
    try {
      const socketUrl = this.getSocketUrl();
      console.log('Connecting to socket at:', socketUrl);
      
      this.socket = io(socketUrl, {
        transports: ['websocket', 'polling'],
        auth: {
          token: this.accessToken
        }
      });

      this.socket.on('connect', () => {
        console.log('Socket connected:', this.socket?.id);
      });

      this.socket.on('disconnect', () => {
        console.log('Socket disconnected');
      });

      this.socket.on('connect_error', (error) => {
        console.error('Socket connection error:', error);
      });
    } catch (error) {
      console.error('Failed to connect to socket server:', error);
    }
  }

  use(databaseId?: string): void {
    this.currentDatabaseId = databaseId;
  }

  setToken(accessToken: string): void {
    this.accessToken = accessToken;
    storage.set(TOKEN_KEYS.ACCESS_TOKEN, accessToken);
  }

  setTokens(accessToken: string, refreshToken: string): void {
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    storage.set(TOKEN_KEYS.ACCESS_TOKEN, accessToken);
    storage.set(TOKEN_KEYS.REFRESH_TOKEN, refreshToken);
  }

  clearTokens(): void {
    this.accessToken = undefined;
    this.refreshToken = undefined;
    storage.remove(TOKEN_KEYS.ACCESS_TOKEN);
    storage.remove(TOKEN_KEYS.REFRESH_TOKEN);
  }

  /**
   * Get socket instance
   */
  getSocket(): Socket | null {
    return this.socket;
  }

  /**
   * Join a channel for listening to events
   */
  joinChannel(databaseId: string | undefined, tableId: string): void {
    if (this.socket) {
      this.socket.emit('join', { database_id: databaseId, table_id: tableId });
    }
  }

  /**
   * Leave a channel
   */
  leaveChannel(databaseId: string | undefined, tableId: string): void {
    if (this.socket) {
      this.socket.emit('leave', { database_id: databaseId, table_id: tableId });
    }
  }

  /**
   * Listen to events on a channel
   */
  on(event: string, callback: (data: any) => void): void {
    if (this.socket) {
      this.socket.on(event, callback);
    }
  }

  /**
   * Remove event listener
   */
  off(event: string, callback?: (data: any) => void): void {
    if (this.socket) {
      if (callback) {
        this.socket.off(event, callback);
      } else {
        this.socket.off(event);
      }
    }
  }

  private loadTokensFromStorage(): void {
    this.accessToken = storage.get(TOKEN_KEYS.ACCESS_TOKEN) || undefined;
    this.refreshToken = storage.get(TOKEN_KEYS.REFRESH_TOKEN) || undefined;
  }

  private async performTokenRefresh(): Promise<boolean> {
    this.isRefreshing = true;
    if (!this.refreshToken) {
      this.isRefreshing = false;
      return false;
    }

    try {
      let query = `refreshToken=${this.refreshToken}`;
      if (this.currentDatabaseId) {
        query += `&database_id=${this.currentDatabaseId}`;
      }
      const url = buildApiUrl(this.baseUrl, '/auth/refresh-token', query);
      const response = await axios.post(url);
      const { data } = response.data as ApiResponse<{ tokens: AuthTokens }>;
      
      if (data?.tokens) {
        this.setTokens(data.tokens.accessToken, data.tokens.refreshToken);
        this.isRefreshing = false;
        return true;
      }
      
      this.clearTokens();
      this.isRefreshing = false;
      return false;
    } catch (error) {
      this.clearTokens();
      this.isRefreshing = false;
      return false;
    }
  }

  public async request<T = any>(
    method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE',
    path: string,
    options: {
      data?: any;
      params?: Record<string, any>;
      config?: AxiosRequestConfig;
    } = {}
  ): Promise<ApiResponse<T>> {
    try {
      const { data, params = {}, config = {} } = options;
      
      if (this.currentDatabaseId) {
        params.database_id = this.currentDatabaseId;
      }
      
      const queryParams = buildQueryParams(params);
      const url = buildApiUrl('', path, queryParams);
      
      const response: AxiosResponse<ApiResponse<T>> = await this.axiosInstance.request({
        method,
        url,
        data,
        ...config,
      });
      
      return response.data;
    } catch (error: any) {
      return {
        error: error.response?.data?.error || error.message || 'Request failed'
      };
    }
  }
}

File: dallosh/servers/dallosh_bot/src/lib/sodular/api/base_client.py
"""
Base Client for Sodular API
Exact Python equivalent of base-client.ts
"""

import asyncio
import json
import re
from typing import Optional, Dict, Any, List, Callable, Literal
import aiohttp
import socketio
from ..types.schema import ApiResponse, AuthTokens
from ..utils import build_query_params, build_api_url, storage, TOKEN_KEYS


class SodularClientConfig:
    """Configuration class for Sodular client"""
    
    def __init__(self, base_url: str, timeout: int = 30000, enable_socket: bool = True):
        self.baseUrl = base_url
        self.timeout = timeout
        self.enableSocket = enable_socket


class BaseClient:
    """Exact Python equivalent of BaseClient class"""
    
    def __init__(self, config: SodularClientConfig):
        self.baseUrl = config.baseUrl
        self.currentDatabaseId: Optional[str] = None
        self.accessToken: Optional[str] = None
        self.refreshToken: Optional[str] = None
        self.isRefreshing = False
        self.refreshPromise: Optional[asyncio.Future] = None
        self.socket: Optional[socketio.AsyncClient] = None
        self.session: Optional[aiohttp.ClientSession] = None
        self.timeout = config.timeout
        self.enableSocket = config.enableSocket
        
        # Create axios equivalent using aiohttp
        self.axiosInstance = self
        self._setup_interceptors()
        self.loadTokensFromStorage()
    
    def _setup_interceptors(self):
        """Setup request/response interceptors equivalent to axios interceptors"""
        # This will be handled in the request method
        pass
    
    async def connect(self) -> Dict[str, Any]:
        """Connect to the server and establish socket connection"""
        print(f"🔌 BaseClient.connect() called for URL: {self.baseUrl}")
        
        try:
            # Test connection with health check
            print("🏥 Testing connection with health check...")
            async with aiohttp.ClientSession() as test_session:
                health_url = f"{self.baseUrl}/health"
                print(f"🔍 Health check URL: {health_url}")
                
                async with test_session.get(health_url) as response:
                    print(f"📊 Health check response status: {response.status}")
                    if response.status != 200:
                        raise Exception(f"Health check failed with status {response.status}")
            
            print("✅ Health check passed")
            
            # Create persistent session
            print("🔧 Creating persistent aiohttp session...")
            self.session = aiohttp.ClientSession(
                timeout=aiohttp.ClientTimeout(total=self.timeout / 1000)
            )
            print("✅ Persistent session created")
            
            # Connect to socket server only if enabled
            if self.enableSocket:
                print("🔌 Setting up socket connection...")
                self.connectSocket()
                print("✅ Socket connection setup completed")
            else:
                print("⚠️ Socket connection disabled")
            
            print("🎉 BaseClient connection successful")
            return {"isReady": True}
            
        except Exception as error:
            print(f"❌ BaseClient connection failed: {error}")
            # Cleanup session if it was created
            if hasattr(self, 'session') and self.session:
                print("🧹 Cleaning up failed session...")
                await self.session.close()
                self.session = None
                print("✅ Session cleanup completed")
            
            return {
                "isReady": False,
                "error": str(error) or "Failed to connect to Sodular backend"
            }
    
    def getSocketUrl(self) -> str:
        """Extract host URL from API base URL for socket connection"""
        try:
            # Remove common API paths
            socket_url = self.baseUrl
            socket_url = re.sub(r'/api/v\d+/?$', '', socket_url)
            socket_url = re.sub(r'/api/?$', '', socket_url)
            socket_url = re.sub(r'/$', '', socket_url)
            return socket_url
        except Exception as error:
            print(f"Failed to parse base URL for socket connection: {error}")
            return self.baseUrl
    
    def connectSocket(self):
        """Connect to socket server"""
        if not self.enableSocket:
            print("⚠️ Socket is disabled, skipping socket connection")
            return
            
        try:
            socket_url = self.getSocketUrl()
            print(f"Connecting to socket at: {socket_url}")
            
            self.socket = socketio.AsyncClient()
            
            @self.socket.event
            async def connect():
                print(f"Socket connected: {self.socket.sid}")
            
            @self.socket.event
            async def disconnect():
                print("Socket disconnected")
            
            @self.socket.event
            async def connect_error(data):
                print(f"Socket connection error: {data}")
            
            # Note: In Python, we'll connect when needed rather than immediately
            # This maintains the same interface as the JavaScript version
            
        except Exception as error:
            print(f"Failed to connect to socket server: {error}")
    
    def use(self, databaseId: Optional[str] = None):
        """Set the current database context"""
        self.currentDatabaseId = databaseId
    
    def setToken(self, accessToken: str):
        """Set the access token"""
        self.accessToken = accessToken
        storage.set(TOKEN_KEYS.ACCESS_TOKEN, accessToken)
    
    def setTokens(self, accessToken: str, refreshToken: str):
        """Set both access and refresh tokens"""
        self.accessToken = accessToken
        self.refreshToken = refreshToken
        storage.set(TOKEN_KEYS.ACCESS_TOKEN, accessToken)
        storage.set(TOKEN_KEYS.REFRESH_TOKEN, refreshToken)
    
    def clearTokens(self):
        """Clear all tokens"""
        self.accessToken = None
        self.refreshToken = None
        storage.remove(TOKEN_KEYS.ACCESS_TOKEN)
        storage.remove(TOKEN_KEYS.REFRESH_TOKEN)
    
    def getSocket(self):
        """Get socket instance"""
        return self.socket if self.enableSocket else None
    
    def joinChannel(self, databaseId: Optional[str], tableId: str):
        """Join a channel for listening to events"""
        if self.socket and self.enableSocket:
            asyncio.create_task(self.socket.emit('join', {
                'database_id': databaseId,
                'table_id': tableId
            }))
    
    def leaveChannel(self, databaseId: Optional[str], tableId: str):
        """Leave a channel"""
        if self.socket and self.enableSocket:
            asyncio.create_task(self.socket.emit('leave', {
                'database_id': databaseId,
                'table_id': tableId
            }))
    
    def on(self, event: str, callback: Callable):
        """Listen to events on a channel"""
        if self.socket and self.enableSocket:
            self.socket.on(event, callback)
    
    def off(self, event: str, callback: Optional[Callable] = None):
        """Remove event listener"""
        if self.socket and self.enableSocket:
            if callback:
                self.socket.off(event, callback)
            else:
                self.socket.off(event)
    
    def loadTokensFromStorage(self):
        """Load tokens from storage"""
        self.accessToken = storage.get(TOKEN_KEYS.ACCESS_TOKEN)
        self.refreshToken = storage.get(TOKEN_KEYS.REFRESH_TOKEN)
    
    async def performTokenRefresh(self) -> bool:
        """Perform token refresh"""
        self.isRefreshing = True
        
        if not self.refreshToken:
            self.isRefreshing = False
            return False
        
        try:
            query = f"refreshToken={self.refreshToken}"
            if self.currentDatabaseId:
                query += f"&database_id={self.currentDatabaseId}"
            
            url = build_api_url(self.baseUrl, '/auth/refresh-token', query)
            
            async with self.session.post(url) as response:
                if response.status == 200:
                    data = await response.json()
                    response_data = data.get('data', {})
                    tokens = response_data.get('tokens')
                    
                    if tokens:
                        self.setTokens(tokens['accessToken'], tokens['refreshToken'])
                        self.isRefreshing = False
                        return True
                    
                    self.clearTokens()
                    self.isRefreshing = False
                    return False
                    
        except Exception:
            self.clearTokens()
            self.isRefreshing = False
            return False
    
    async def request(self, method: Literal['GET', 'POST', 'PUT', 'PATCH', 'DELETE'], path: str, options: Dict[str, Any] = None) -> ApiResponse:
        """Make HTTP request - EXACTLY like JavaScript"""
        if options is None:
            options = {}
        
        if not self.session:
            return {"error": "Client not connected. Call connect() first."}
        
        try:
            data = options.get('data')
            params = options.get('params', {})
            config = options.get('config', {})
            
            # Add database context if available - EXACTLY like JavaScript
            if self.currentDatabaseId:
                params['database_id'] = self.currentDatabaseId
            
            # Build URL with full base URL since we don't have axios baseURL
            query_params = build_query_params(params)
            url = build_api_url(self.baseUrl, path, query_params)
            
            # Set up headers with authentication
            headers = {}
            if self.accessToken:
                headers['Authorization'] = f'Bearer {self.accessToken}'
            
                           # Make request using aiohttp - EXACTLY like JavaScript
                async with self.session.request(
                   method=method,
                   url=url,
                   json=data,
                   headers=headers,
                   **config
                ) as response:
                   # Check for successful status codes (2xx range)
                    if 200 <= response.status < 300:
                       return await response.json()
                    else:
                       return {"error": f"Request failed with status {response.status}"}
                    
        except aiohttp.ClientError as error:
            return {"error": f"Network error: {str(error)}"}
        except Exception as error:
            return {"error": str(error) or "Request failed"}
    
    async def close(self):
        """Close the client and cleanup resources"""
        try:
            if self.socket and self.enableSocket:
                await self.socket.disconnect()
                self.socket = None
        except Exception as e:
            print(f"Error disconnecting socket: {e}")
        
        try:
            if self.session:
                await self.session.close()
                self.session = None
        except Exception as e:
            print(f"Error closing session: {e}")
        
        # Clear tokens
        self.clearTokens()
